src/main/scala/com/cognite/spark/datasource/AssetsRelationV1.scala

```diff
@@ -1,23 +1,46 @@
 package com.cognite.spark.datasource
 
 import cats.effect.IO
-import com.cognite.sdk.scala.v1.{Asset, AssetCreate, AssetUpdate, GenericClient}
+import com.cognite.sdk.scala.v1.{Asset, AssetCreate, AssetUpdate, AssetsFilter, GenericClient}
 import com.cognite.sdk.scala.v1.resources.Assets
+import com.cognite.sdk.scala.common
 import org.apache.spark.sql.{Row, SQLContext}
-import org.apache.spark.sql.sources.InsertableRelation
+import org.apache.spark.sql.sources.{Filter, InsertableRelation}
 import io.circe.generic.auto._
 import com.cognite.spark.datasource.SparkSchemaHelper._
 import com.softwaremill.sttp.Uri
 import com.softwaremill.sttp._
 import org.apache.spark.sql.types._
 import cats.implicits._
+import PushdownUtilities._
 import AssetsRelation.fieldDecoder
 import com.cognite.sdk.scala.common.CdpApiException
 
 class AssetsRelationV1(config: RelationConfig)(val sqlContext: SQLContext)
     extends SdkV1Relation[Asset, Assets[IO], AssetsItem](config, "assets")
     with InsertableRelation {
 
+  override def getReaderIO(filters: Array[Filter])(
+      client: GenericClient[IO, Nothing],
+      cursor: Option[String],
+      limit: Option[Long]): Seq[IO[common.ItemsWithCursor[Asset]]] = {
+    val fieldNames = Array("name", "source")
+    val pushdownFilterExpression = toPushdownFilterExpression(filters)
+    val getAll = shouldGetAll(pushdownFilterExpression, fieldNames)
+    val params = pushdownToParameters(pushdownFilterExpression)
+
+    val pushdownFilters = if (params.isEmpty || getAll) {
+      Seq(AssetsFilter())
+    } else {
+      params.map(assetsFilterFromMap)
+    }
+
+    pushdownFilters.map(f => client.assets.filterWithCursor(f, cursor, limit))
```

Comment:
I added a commit with a suggestion showing how to use `parTraverse` instead, that would fetch those in parallel.

Reply:
Nice, thanks!

Reply:
Looks like we're running into problems serializing `ContextShift`. See any way around it?
https://github.com/typelevel/cats-effect/issues/587

Reply:
We use `ContextShift` elsewhere, so it should be possible. Maybe it needs to be marked as transient and/or lazy?

Reply:
I guess that's why we define it inline with the method in other places, but it would be better if we could have it as an instance variable.